src/phase0/state_transition/epoch_processing.rs

```diff
@@ -177,8 +178,29 @@ pub fn process_slashings<
         MAX_VALIDATORS_PER_COMMITTEE,
         PENDING_ATTESTATIONS_BOUND,
     >,
-    _context: &Context,
+    context: &Context,
 ) -> Result<(), Error> {
+    let epoch = get_current_epoch(state, context);
+    let total_balance = get_total_active_balance(state, context)?;
+    let adjusted_total_slashing_balance = Gwei::min(
+        state.slashings.iter().sum::<Gwei>() * context.proportional_slashing_multiplier,
+        total_balance,
+    );
+    let validators = state.validators.clone();
```

Comment:
So, I tried to avoid using `clone()`, but I ran into issues if there's any guidance on better alternatives. 

For example:
```
let validators = state.validators.iter().enumerate();
for (index, validator) in validators {
   ...
   decrease_balance(state, index, penalty);
}
```
When I try to iterate through the data, I get a _"cannot borrow `*state` as mutable because it is also borrowed as immutable"_ error. My understanding is that when this local `validators` variable is created, it borrows state as immutable; thus, when the `for` loop tries to call `decrease_balance` with `state`, it's trying to mutate it -- so, Rust throws an error. 

Basically, since `state.validators` won't be altered during the loop, I'd like to be able to just read it via a reference or something while `decrease_balance` is mutating `state.balances` if that makes sense. I briefly looked into memory-related things like `Cell`, `RefCell`, `RwLock`, `Arc`, etc., but idk if those are needed or how to use them properly.

Reply:
right, this comes down to how rust handles borrows, so we will have to be a little careful about how we do this.

when you say `state.validators.iter()`, you get back a thing that essentially has an immutable reference inside it (generated from the "broader" mutable reference from `state: &mut T`)

and the borrow rule is basically "1 writer xor many readers" -- so if we hand out the immutable ref, we cannot use the mutable ref down below in `decrease_balance` as the immutable ref that is live stays live until the end of its enclosing scope.

there are a few ways to handle this particular problem (e.g. you could make a separate list of the balance deltas to apply and then basically drop the validators ref and *then* apply the deltas) -- one that i think sticks closely to the spec is to just iterate over indices and take advantage of the fact that the `validator` ref is not used in `decrease_balance`

refer to this block: https://github.com/ralexstokes/ethereum_consensus/blob/main/src/phase0/state_transition/genesis.rs#L87-L99

for an example, although you should be able to just say `let validator = &state.validators[i];`. this ends up being fine as this ref for the validator is dropped before we "reborrow" the mutable ref for state in `decrease_balance`

Reply:
as a side note, the `validators` field is one of the larger ones in the state so we definitely don't want to be copying it where we can avoid it

Reply:
Ahh okay, I think I get it now. By using `validators` only within the `for` loop's scope (and not prior to the loop executing), the immutable reference to `state` is last used prior to `decrease_balance` borrowing `state` as mutable...so, then there's no issue with multiple immutable/mutable references at the same time.